controllers/reviewController.ts

```diff
@@ -0,0 +1,129 @@
+import {Request, Response} from 'express';
+import verifyTokenAndRetrieveUser from "../service/userService.js";
+import ProductModel from "../model/product.js";
+import ReviewModel from "../model/review.js";
+import OrderModel from "../model/order.js";
+
+/**
+ * The function adds a new review for a product if the user is authenticated and has bought the product.
+ * @param req Request, productId: String, userId: String, reviewText: String, reviewTitle: String, starRating: Number
+ * @param res Response, Review: Object
+ * @return {Object} The newly added review.
+ * @throws {400} If there's an error while adding the review.
+ */
+export const addReview = async (req: Request, res: Response) => {
+    try {
+        const {productId, userId, reviewText, reviewTitle, starRating} = req.body,
+            token = req.headers.authorization?.split(' ')[1];
+
+        if (!(productId && userId && reviewText && reviewTitle && starRating))
+            return res.status(404).json({error: 'Missing fields'});
+
+        const user = await verifyTokenAndRetrieveUser(token, userId),
+            product = await ProductModel.findOne({id: productId});
+        if (!product) {
+            return res.status(404).json({error: 'Product not found'});
+        }
+        if (!user) {
+            return res.status(404).json({error: 'User not found'});
+        }
+
+        const orders = await OrderModel.find({userId: userId.trim()});
+
+        let ownsProduct = orders.some(order => {
+            return order.products.some(product => {
+                if (product.productId === productId) {
+                    return true;
+                }
+            })
+        })
+
+        if (!ownsProduct) {
+            res.status(404).json({error: 'You cannot review a product you have not bought'});
+            return;
+        }
+
+        const foundReview = await ReviewModel.findOne({userId, productId: productId});
+
+        if (foundReview)
+            return res.status(400).json({error: 'You have already reviewed this product'});
+
+        const currentDate = new Date(),
+            newReview = await ReviewModel.create({
+                userId,
+                productId,
+                starRating,
+                reviewTitle,
+                reviewText,
+                date: currentDate
+            });
+        await newReview.save();
```

Comment:
poti sa faci un orderService in care acolo sa pui toata logica , aici doar sa apelezi functia cu parametrii